lib/MetaCPAN/Model/Tarball.pm

```diff
@@ -0,0 +1,167 @@
+    coerce   => 1,
+);
+
+has files => (
+    is       => 'ro',
+    isa      => ArrayRef,
+    init_arg => undef,
+    lazy     => 1,
+    builder  => '_build_files',
+);
+
+has date => (
+    is  => 'rw',
+    isa => 'DateTime',
+);
+
+has index => ( is => 'rw', );
+
+has author => (
+    is  => 'rw',
+    isa => 'Str',
+);
+
+has name => (
+    is  => 'rw',
+    isa => 'Str',
+);
+
+has metadata => ( is => 'rw', );
+
+has status => (
+    is  => 'rw',
+    isa => 'HashRef',
+);
+
+has distribution => (
+    is  => 'rw',
+    isa => 'Str',
+);
+
+has version => (
+    is  => 'rw',
+    isa => 'Str',
+);
+
+has maturity => (
+    is  => 'rw',
+    isa => 'Str',
+);
+
+has release => (
+    is  => 'rw',
+    isa => 'HashRef',
+);
+
+has tmpdir => (
+    is  => 'rw',
+    isa => Dir,
+);
+
+has bulk => ( is => 'rw', );
+
+sub _build_files {
+    my $self = shift;
+    log_info {'Processing ', $self->tarball};
+
+    my $archive = Archive::Any->new( $self->tarball );
+
+    log_error {"$self->tarball is being impolite"} if $archive->is_impolite;
+
+    log_error {"$self->tarball is being naughty"} if $archive->is_naughty;
+
+    log_debug {'Extracting archive to filesystem'};
+    $archive->extract( $self->tmpdir );
+
+    my @files;
+    log_debug { 'Indexing ', scalar $archive->files, 'files' };
+    my $file_set = $self->index->type('file');
+
+    File::Find::find(
+        sub {
+            my $child
+                = -d $File::Find::name
+                ? dir($File::Find::name)
+                : file($File::Find::name);
+            my $relative = $child->relative( $self->tmpdir );
+            my $stat     = do {
+                my $s = $child->stat;
+                +{ map { $_ => $s->$_ } qw(mode uid gid size mtime) };
+            };
+            return if _is_broken_symlink($File::Find::name);
+            return if _is_broken_file($File::Find::name);
+            return if ( $relative eq q{.} );
+            ( my $fpath = "$relative" ) =~ s/^.*?\///;
+            my $fname = $fpath;
+            $child->is_dir
+                ? $fname =~ s/^(.*\/)?(.+?)\/?$/$2/
+                : $fname =~ s/.*\///;
+            $fpath = q{} if $relative !~ /\// && !$archive->is_impolite;
+
+            my $file = $file_set->new_document(
+                Dlog_trace {"adding file $_"} +{
+                    metadata     => $self->metadata,
+                    name         => $fname,
+                    directory    => $child->is_dir,
+                    release      => $self->name,
+                    date         => $self->date,
+                    distribution => $self->distribution,
+                    author       => $self->author,
+                    local_path   => $child,
+                    path         => $fpath,
+                    version      => $self->version,
+                    stat         => $self->status,
+                    maturity     => $self->maturity,
+                    status       => $self->release->status,
```

Comment:
If you alpha sort the hash keys, this becomes easier to read and duplicate keys become easier to spot.